internals/overlord/logstate/gatherer.go

```diff
@@ -0,0 +1,190 @@
+// Copyright (c) 2023 Canonical Ltd
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License version 3 as
+// published by the Free Software Foundation.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package logstate
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/canonical/pebble/internals/logger"
+	"github.com/canonical/pebble/internals/plan"
+	"github.com/canonical/pebble/internals/servicelog"
+)
+
+const parserSize = 4 * 1024
+
+// logGatherer is responsible for collecting service logs from a bunch of
+// services, and sending them to its logClient.
+// One logGatherer will run per log target. Its loop() method should be run
+// in its own goroutine. It can be stopped in a separate goroutine by calling
+// the stop() method.
+// The logGatherer will "flush" the client:
+//   - on a regular cadence (e.g. every 1 second)
+//   - when it is told to shut down.
+//
+// Its client may also flush itself when its internal buffer reaches a certain
+// size.
+type logGatherer struct {
+	target     *plan.LogTarget
+	tickPeriod time.Duration
+	cancel     chan struct{}
+	client     logClient
+
+	iterators  map[string]servicelog.Iterator
+	parsers    map[string]*servicelog.Parser
+	itLock     sync.Mutex
+	notifyChan chan bool
+}
+
+func newLogGatherer(target *plan.LogTarget) (*logGatherer, error) {
+	client, err := newLogClient(target)
+	if err != nil {
+		return nil, fmt.Errorf("could not create log client: %w", err)
+	}
+
+	return &logGatherer{
+		target:     target,
+		tickPeriod: 1 * time.Second,
+		cancel:     make(chan struct{}),
+		client:     client,
+		notifyChan: make(chan bool, 1),
+	}, nil
+}
+
+func (g *logGatherer) addIterator(serviceName string, iterator servicelog.Iterator, parser *servicelog.Parser) {
+	g.itLock.Lock()
+	defer g.itLock.Unlock()
+
+	iterator.Notify(g.notifyChan)
```

Comment:
Given that the iteration is surviving across multiple plans while gatherers get removed, how do we clean up notifiers so it doesn't become a leak?